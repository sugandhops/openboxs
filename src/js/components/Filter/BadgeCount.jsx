import React from 'react';

import PropTypes from 'prop-types';

const BadgeCount = ({ count }) => (
  <span className="d-flex justify-content-center align-items-center badge-count">
    {count}
  </span>
);

export default BadgeCount;

BadgeCount.propTypes = {
  count: PropTypes.number.isRequired,
};
