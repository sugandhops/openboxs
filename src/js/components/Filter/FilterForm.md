## FilterForm documentation
* `<FilterForm>`
#### Available props:
| Name                   | Type              | Values                                            | Required | Description                                                                                                                             |   
|------------------------|-------------------|---------------------------------------------------|----------|-----------------------------------------------------------------------------------------------------------------------------------------|
| filterFields           | object of objects | check example below                               | yes      | The same type as we use for<br/>normal forms (check below)                                                                              | 
| updateFilterParams     | function          |                                                   | yes      | Action to trigger after submiting filter form or assigning default values (e.g. fetching data with values from filter's form as params) | 
| searchFieldPlaceholder | string            | by default: 'Search'                              | no       | All filter forms are supposed to have search field, so provide placeholder for that field as props                                      |
| formProps              | object            | check example below                               | no       | Form props for some fields' needs, e.g. **getDynamicAttr**                                                                              | 
| searchFieldId          | string            | id of searchField input (by default 'searchTerm') | no       | Id of search field - check what backend expects for searchTerm checking (could be 'searchTerm', 'name' etc.)                            | 
| defaultValues          | object            |                                                   | no       | Default values for filter form                                                                                                          | 
| allowEmptySubmit       | boolean           | true/false (false by default)                     | no       | Prop to determine if Search button should be enabled/disabled when empty values only                                                    | 
| hidden                 | boolean           | true/false (true by default)                      | no       | Hide filters by default                                                                                                                 | 
#### Example:
````md
const filterFields = {
    address: {
        type: TextField,
        label: 'address.description.label',
        defaultMessage: 'Address',
    },
    destination: {
        type: SelectField,
        label: 'react.inboundReturns.destination.label',
        defaultMessage: 'Destination',
        attributes: {
            required: true,
            async: true,
            showValueTooltip: true,
            openOnClick: false,
            autoload: false,
            cache: false,
            options: [],
            filterOptions: options => options,
        },
        getDynamicAttr: props => ({
            loadOptions: props.debouncedDestinationLocationsFetch,
        }),
    },
    date: {
        type: DateField,
        label: 'react.invoice.invoiceDate.label',
        defaultMessage: 'Date',
        attributes: {
            dateFormat: 'MM/DD/YYYY',
        },
    },
}

const debouncedOriginLocationsFetch =
    debounceLocationsFetch(
        debounceTime,
        minSearchLength,
        [], // activityCodes
        false, // fetchAll
        false, // withOrgCode
        true, // withTypeDescription
        true, // isReturnOrder
    );   

<FilterForm
    onSubmit={values => console.log(values)}
    filterFields={filterFields}
    searchFieldPlaceholder="Search...."
    formProps={
        {
            debouncedOriginLocationsFetch: debouncedOriginLocationsFetch, 
        }
    }
/>
````


<br>
<br>

* `<FilterVisibilityToggler>`
#### Available props:
| Name             | Type     | Values                              | Required | Description                                                                                                           |   
|------------------|----------|-------------------------------------|----------|-----------------------------------------------------------------------------------------------------------------------|
| filtersHidden    | boolean  | true/false                          | yes      | State determining if filters are shown to display correct div with either Hide or Show filters label and correct icon | 
| amountFilled     | number   | amountFilled from `<FilterForm>`    | yes      | Calculated amount of filled fields provided further to `<BadgeCount>`                                                 | 
| setFiltersHidden | function | setFiltersHidden from `<FilterForm>` | yes      | Set state function to hide/show filters                                                                               |

<br>
<br>

* `<BadgeCount>`
#### Available props:
| Name  | Type   | Values | Required | Description                            |   
|-------|--------|--------|----------|----------------------------------------|
| count | number |        | yes      | Simple number to display inside circle | 

