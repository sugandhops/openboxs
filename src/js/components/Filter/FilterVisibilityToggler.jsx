import React from 'react';

import PropTypes from 'prop-types';
import { RiArrowDownSLine, RiArrowUpSLine } from 'react-icons/ri';

import BadgeCount from 'components/Filter/BadgeCount';
import Translate from 'utils/Translate';

const FilterVisibilityToggler = ({ filtersHidden, amountFilled, setFiltersHidden }) => {
  const labelAttributes = filtersHidden
    ? { id: 'react.button.showFilters.label', defaultMessage: 'Show Filters' }
    : { id: 'react.button.hideFilters.label', defaultMessage: 'Hide Filters' };

  return (
    <button
      className="d-flex align-items-center hide-filters-button"
      type="button"
      onClick={() => setFiltersHidden(!filtersHidden)}
    >
      <span className="hide-filters-label">
        <Translate {...labelAttributes} />
      </span>
      { amountFilled > 0 && <BadgeCount count={amountFilled} />}
      { filtersHidden ? <RiArrowDownSLine /> : <RiArrowUpSLine /> }
    </button>
  );
};

export default FilterVisibilityToggler;

FilterVisibilityToggler.propTypes = {
  filtersHidden: PropTypes.bool.isRequired,
  amountFilled: PropTypes.number.isRequired,
  setFiltersHidden: PropTypes.func.isRequired,
};
