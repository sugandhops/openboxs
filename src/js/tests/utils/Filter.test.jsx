import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Filter from 'utils/Filter';

import '@testing-library/jest-dom';

let filter;
let filterWithValue;
let onChangeFunc;

describe('filter', () => {
  beforeEach(() => {
    onChangeFunc = jest.fn();
    filter = (<Filter onChange={onChangeFunc} />);
    filterWithValue = (<Filter onChange={onChangeFunc} filter={{ value: 'testValue' }} />);
  });

  it('should match snapshot', () => {
    const renderedFilter = renderer.create(filter);
    expect(renderedFilter.toJSON())
      .toMatchSnapshot();
  });

  it('should render properly', () => {
    render(filter);
    expect(screen.getByTestId('filter-input'))
      .toBeInTheDocument();
  });

  it('should call function when data changes', () => {
    render(filter);
    expect(onChangeFunc)
      .toHaveBeenCalledTimes(0);
    fireEvent.change(screen.getByTestId('filter-input'), { target: { value: 'a' } });
    expect(onChangeFunc)
      .toHaveBeenCalledTimes(1);
    expect(onChangeFunc)
      .toHaveBeenCalledWith('a');
  });

  it('should have starting value when filter.value is present', () => {
    render(filterWithValue);
    expect(screen.getByTestId('filter-input'))
      .toHaveAttribute('value', 'testValue');
  });
});
