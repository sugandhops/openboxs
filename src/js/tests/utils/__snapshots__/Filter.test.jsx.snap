// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter should match snapshot 1`] = `
<input
  data-testid="filter-input"
  onChange={[Function]}
  placeholder="Search"
  style={
    {
      "width": "100%",
    }
  }
  value=""
/>
`;
